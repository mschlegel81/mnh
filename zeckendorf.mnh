memoized fib(i> 0)->fib(i-1)+fib(i-2);
         fib(i<=0):=1;
inv(F:expression,n  )->inv(F,n,0);
inv(F:expression,n,i)->(F(i))>=n ? i : inv(F,n,i+1);

encode(F,n     )->replace('#'&encode(F,n,inv(F,n)),["#0","#"],["#",""]);
encode(F,n,i>=0)->(F(i))<=n ? 1 & encode(F,n-(F(i)),i-1)
                            : 0 & encode(F,n       ,i-1);
encode(F,n,i)->"";

decode(F,s:string)->decode(F,[length(s)-1..0]%(copy(s,[0..length(s)-1],1)="1"));
decode(F,K:intList)->each(k,K,+,F(k));
decode(F,[])->0;

FIB:={fib($i)};
POT:={2^$i};

print(each(i,[0..100],&"\n"&,encode(POT,i)&"\t"&
                             encode(FIB,i)&"\t"&
                             decode(POT,encode(FIB,i))&"\t"&
                             decode(FIB,encode(POT,i))));
