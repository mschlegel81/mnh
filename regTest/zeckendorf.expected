 decimal   binary Zeckendorf
       1        1          1
       2       10         10
       3       11        100
       4      100        101
       5      101       1000
      17    10001     100101
     100  1100100 1000010100
     255 11111111 1000010000

      bits rel. overhead
 1    1.00        1.0000
 2    2.50        1.2500
 3    3.75        1.2500
 4    5.38        1.3438
 5    6.69        1.3375
 6    8.22        1.3698
 7    9.61        1.3728
 8   11.05        1.3818
 9   12.53        1.3919
10   13.88        1.3881
11   15.44        1.4037
12   16.74        1.3949
